### DISCUSSION TITLE
Sorting order is different from a Luhmann slipbox

### DISCUSSION THREAD

**Discussion by ManuelRodriguez331** [zim-desktop-wiki/zim-desktop-wiki#2128] [General]
The Zimwiki software is using the sorting order of the underlying file system. Unfortunately, this will sort top level entries created with alphanumerical Luhmann style numbering system to the wrong position. Example:

current sorting order is
```
“12a hello”
“12 hello”
```

needed sorting order is
```
“12 hello”
“12a hello”
```

The problem has to do with the length of the numbering. The resulting sad situation is that top level entries in a Luhmann zettelkasten are randomly distributed below low level entries which is very confusing for beginners in personal knowledge systems.

**Comment by introt** [comment]
Postfixing a `0` does the trick, albeit it looks a bit weird:

```
"120 hello"
"12a hello"
```

Though given that the Luhmann identifiers have hierarchy, I'd use sub-pages for low level entries: `12 hello:12a hello` (if limited to using the "original" numbering scheme).

Related issue: #32

**Reply by ManuelRodriguez331** [reply]
Perhaps it makes sense to introduce first what a Luhmann Zettelkasten is about. It is some sort of 1d array which holds a mindmap. The alphanumerical numbering system of the title section is very important to maintain the structure.

> Postfixing a 0 does the trick, albeit it looks a bit weird:

It is sorted correctly in the file manager, but it remains in the wrong order in the left pane of zimwiki. In addition, its not practical to fill all the missing zeros because the hierarchy goes up to 10 levels deep. This would produce an entry like 1200000000 which is very confusing.

Here is a code snippet which works better:

```
#!/usr/bin/python3
rawlist={
  "12a2a hello",
  "12a2 hello",
  "12a hello",
  "12b hello",
  "12a1 hello",
  "12 xyz",
}
print(rawlist)
print(sorted(rawlist))

# output is:
{'12a1 hello', '12a2a hello', '12b hello', '12a2 hello', '12 xyz', '12a hello'}
['12 xyz', '12a hello', '12a1 hello', '12a2 hello', '12a2a hello', '12b hello']
```

It seems, that the builtin “sorted()” function of python will produce the correct Luhmann style sorting order as default without further improvements. The only question is how to insert the code snippet into the zimwiki github project?

**Comment by shawnhcorey** [comment]
On 2022-08-16 14:39, ManuelRodriguez331 wrote:
> The only question is how to insert the code snippet into the zimwiki 
> github project?

Clone the repository, make the changes, test the changes, make a pull 
request to the owner of the original repository.

**Reply by ManuelRodriguez331** [reply]
**Sorting order can be changed with the utils.py file**

The utils.py file has from line 116 and ongoing a subroutine which is responsible for the sorting order in the left pane of the zim wiki. The self understanding of the files in the pane is, that these list is called an index.

It is important to know, that zimwiki creates an intermediate layer between the GUI and the filesystem which is realized with a sqlite3 database. That means, the filenames are read in during startup and then converted into an in-memory SQL database. The sorting order in this database is realized with the ` natural_sort_key()` function.

The unsolved problem is, what this function is doing exactly. It takes for a string `“text_1.2.3_unicode_αβγ_žžž”` as input and converts it into a hex value. This can be seen by adding some print statement into the sourcecode to get a verbose output to the command line. But it remains a mystery how to realize a new sorting order.

It seems, that the 50k lines of code of the project have grown up over a longer period ago and it is hard to identify the reason why the system is in the current state. In a previous post here in the forum #1442 the sorting order was mentioned in the past.

**Comment by jaap-karssenberg** [comment]
The sorting function is doing two things: a) ensure we respect unicode /
locale specific alphabetical sorting (e.g. AaBa instead of ABab)  and b)
try to get numbers right (e.g. 1 2 10 instead of 1 10 2). The hex value is
needed to encode the resulting sorting key in ascii again so it is safe to
use in the index tables without further unicode / locale headaches.

Hope this helps.

--- Jaap

**Comment by shawnhcorey** [comment]
On 2022-08-15 02:04, ManuelRodriguez331 wrote:
>
> The Zimwiki software is using the sorting order of the underlying file 
> system. Message ID: 
> ***@***.***>
>
Does it? Or does it use locale?

If it uses the system sort, it will use locale. On Linux, to disable 
this and get an ASCII sort, use:

     LC_ALL=C zim

Of course, this will sort all capital letters before all lower case 
ones, so it's not a perfect solution.
